### PR TITLE
add `first`, `last` methods with `n` for gdf

### LINKED ISSUE #2993: `first(gdf::GroupedDataFrame, n::Int)` should give a `GroupedDataFrame`?
`first(df::DataFrame, n)` gives a `DataFrame` with the first `n` rows (and doesn't error if `n > nrow(df)`, matching Base). It would be nice if `first(gdf, n)` gave the first `n` groups (without erroring if `n` is more than the number of groups). Currently, it gives a `Vector{Any}`. However, one can do `gdf[1:min(n, end)]` instead.

I suppose this might be breaking, in which case it would be a request for DataFrames 2.0.

### PR DESCRIPTION
**Pull request by ericphanson**
closes #2993

### CODE REVIEW

File: src/groupeddataframe/groupeddataframe.jl [hunk 1]
```diff
@@ -429,8 +429,16 @@ end
 Base.axes(gd::GroupedDataFrame, i::Integer) = Base.OneTo(size(gd, i))
 
 Base.first(gd::GroupedDataFrame) = gd[1]
-Base.last(gd::GroupedDataFrame) = gd[end]
+function Base.first(gd::GroupedDataFrame, n::Integer)
+    n < 0 && throw(ArgumentError("Number of elements must be nonnegative"))
+    return @inbounds gd[1:min(n, end)]
```
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:434]: I would remove the `@inbounds` as the time saved by it is negligible and without it the code is safer:
```suggestion
    return gd[1:min(n, end)]
```

File: src/groupeddataframe/groupeddataframe.jl [hunk 2]
```diff
@@ -429,8 +429,16 @@ end
...
+end
 
+Base.last(gd::GroupedDataFrame) = gd[end]
+function Base.last(gd::GroupedDataFrame, n::Integer)
+    n < 0 && throw(ArgumentError("Number of elements must be nonnegative"))
+    return @inbounds gd[max(1, end - n + 1):end]
```
Comment (bkamins) [inline src/groupeddataframe/groupeddataframe.jl:440]:
```suggestion
    return gd[max(1, end - n + 1):end]
```

**Review by bkamins — APPROVED** [review]

**Comment by bkamins** [comment]
The only reservation I have is that we accept the following:
```
julia> x = [5, 6, 7]
3-element Vector{Int64}:
 5
 6
 7

julia> first(x, true)
1-element Vector{Int64}:
 5

julia> first(x, false)
Int64[]
```

I on purpose show how Base Julia works. The question is if we want to accept `Bool`. But maybe it is OK. What do you think?

CC @nalimilan

**Comment by bkamins** [comment]
Ah - and can you maybe add a NEWS.md entry please?

Something like:
```
# DataFrames.jl v1.4 Release Notes

## New functionalities

* `first` and `last` for `GroupedDataFrame` now support passing number of elements to get
  ([#3006](https://github.com/JuliaData/DataFrames.jl/issues/3006))
```

**Comment by ericphanson** [comment]
Updated!

re-inbounds, I was a bit hesitant too; I looked at the Base abstract vector implementation and copied it from there, and convinced myself it was indeed safe. But agreed it doesn’t get us much here.

**Comment by bkamins** [comment]
@nalimilan - I am OK to merge this, but please have a look (especially if we want to allow `Bool` as `n` like in Base Julia)

**Review by nalimilan — APPROVED** [review]
I don't have a strong opinion regarding whether we should accept `Bool` or not.

**Comment by bkamins** [comment]
OK - let us stay consistent with Base Julia then.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**